psi4/src/psi4/libmints/oeprop.cc

```diff
@@ -1784,6 +1788,445 @@ PopulationAnalysisCalc::compute_lowdin_charges(bool print_output) {
+    //wfn_->set_atomic_point_charges(mpole);
+
+    wfn_->set_array_variable("MBIS_CHARGES", mpole);
+    wfn_->set_array_variable("MBIS_DIPOLES", dpole);
+    wfn_->set_array_variable("MBIS_QUADRUPOLES", qpole);
+    wfn_->set_array_variable("MBIS_OCTUPOLES", opole);
+}
+
+int PopulationAnalysisCalc::get_nai(int z, int m) {
+    //MBIS is not supported for elements with atomic number greater than 36
+    if (z <= 2) {
+        return z;
+    } else if (z <= 10) {
+        if (m == 1) {
+            return 2;
+        } else {
+            return z - 2;
+        }
+    } else if (z <= 18) {
+        if (m == 1) {
+            return 2;
+        } else if (m == 2) {
+            return 8;
+        } else {
+            return z - 10;
+        }
+    } else {
+        if (m == 1) {
+            return 2;
+        } else if (m == 2) {
+            return 8;
+        } else if (m == 3) {
+            return 8;
+        } else {
+            return z - 18;
+        }
+    }
+}
+
+double PopulationAnalysisCalc::rho_ai_0(double n, double sigma, double distance) {
+    return n * exp(-distance / sigma) / (pow(sigma, 3) * 8 * M_PI);
+}
+
+std::tuple<SharedMatrix, SharedMatrix, SharedMatrix, SharedMatrix>
+PopulationAnalysisCalc::compute_mbis_multipoles(bool print_output) {
+
+    if (print_output) outfile->Printf("  ==> Computing MBIS Charges <==\n\n");
+    timer_on("MBIS");
+    
+    // => Setup 1RDM on DFTGrid <= //
+
+    Options& options = Process::environment.options;
+    const int MAX_ITER = options.get_int("MBIS_MAXITER");
+    const double CONV = options.get_double("MBIS_CONVERGENCE");
+
+    std::shared_ptr<Molecule> mol = basisset_->molecule();
+    std::shared_ptr<DFTGrid> grid = std::make_shared<DFTGrid>(mol, basisset_, options);
+
+    int nbf = basisset_->nbf();
+    int num_atoms = mol->natom();
+    int total_points = grid->npoints();
+    auto apcs = std::make_shared<std::vector<double>>(num_atoms);
+
+    SharedMatrix Da = wfn_->Da_subset("AO");
+    SharedMatrix Db;
+    std::shared_ptr<PointFunctions> point_func(nullptr);
+
+    if (same_dens_) {
+        Db = Da->clone();
+        point_func = std::make_shared<RKSFunctions>(basisset_, total_points, nbf);
+        point_func->set_pointers(Da);
+    } else {
+        Db = wfn_->Db_subset("AO");
+        point_func = std::make_shared<UKSFunctions>(basisset_, total_points, nbf);
+        point_func->set_pointers(Da, Db);
+    }
+
+    std::vector<std::shared_ptr<BlockOPoints>> blocks = grid->blocks();
+    int num_blocks = blocks.size();
+    int running_points = 0;
+
+    // Coordinates, weights, and rho (molecular electron density) at each grid point
+    std::vector<double> x_points(total_points, 0.0);
+    std::vector<double> y_points(total_points, 0.0);
+    std::vector<double> z_points(total_points, 0.0);
+    std::vector<double> weights(total_points, 0.0);
+    std::vector<double> rho(total_points, 0.0);
+
+    for (int b = 0; b < num_blocks; b++) {
```

Comment:
won't change fn a bit. but since `num_blocks` only used once, probably doesn't need to be separate var.